src/SA/ldapsearch/entry.c

```diff
@@ -0,0 +1,398 @@
+DECLSPEC_IMPORT VOID WINAPI WLDAP32$ldap_memfree(PCHAR);
+
+DECLSPEC_IMPORT ULONG WINAPI WLDAP32$ldap_unbind(LDAP*);
+DECLSPEC_IMPORT ULONG WINAPI WLDAP32$ldap_unbind_s(LDAP*);
+DECLSPEC_IMPORT ULONG WINAPI WLDAP32$ldap_msgfree(LDAPMessage*);
+
+DECLSPEC_IMPORT int __cdecl MSVCRT$strcmp(const char *_Str1,const char *_Str2);
+DECLSPEC_IMPORT PCHAR WINAPI MSVCRT$strstr(const char *haystack, const char *needle);
+
+#define TICKSTO1970         0x019db1ded53e8000LL
+#define TICKSTO1980         0x01a8e79fe1d58000LL
+#define TICKSPERSEC         10000000
+#ifdef _WIN64 //can't use builtin function for extended divition on x86
+DWORD GetTimeInSeconds(VOID)
+{
+    LARGE_INTEGER Time;
+    FILETIME FileTime;
+    DWORD dwSeconds;
+    int junk = 0;
+    lldiv_t test;
+    KERNEL32$GetSystemTimeAsFileTime(&FileTime);
+    Time.u.LowPart = FileTime.dwLowDateTime;
+    Time.u.HighPart = FileTime.dwHighDateTime;
+    Time.QuadPart = Time.QuadPart - TICKSTO1970;
+    Time.QuadPart = (uint64_t)Time.QuadPart / (uint64_t)TICKSPERSEC;
+    dwSeconds = Time.u.LowPart;
+
+    return dwSeconds;
+}
+#endif
+VOID ReturnDateTime(DWORD dwSeconds)
+{
+    LARGE_INTEGER Time;
+    FILETIME FileTime;
+    SYSTEMTIME SystemTime;
+    PCHAR DateBuffer[80];
+    PCHAR TimeBuffer[80];
+
+    Time.QuadPart = ((LONGLONG)dwSeconds * TICKSPERSEC) + TICKSTO1970;
+    
+    FileTime.dwLowDateTime = Time.u.LowPart;
+    FileTime.dwHighDateTime = Time.u.HighPart;
+    KERNEL32$FileTimeToLocalFileTime(&FileTime, &FileTime);
+    KERNEL32$FileTimeToSystemTime(&FileTime, &SystemTime);
+
+    KERNEL32$GetDateFormatW(LOCALE_USER_DEFAULT,
+                   DATE_SHORTDATE,
+                   &SystemTime,
+                   NULL,
+                   DateBuffer,
+                   ARRAYSIZE(DateBuffer));
+
+    KERNEL32$GetDateFormatW(LOCALE_USER_DEFAULT,
+                   TIME_NOSECONDS,
+                   &SystemTime,
+                   NULL,
+                   TimeBuffer,
+                   ARRAYSIZE(TimeBuffer));
+
+    internal_printf("%S %S", DateBuffer, TimeBuffer);
+}
+
+PCHAR GetDomainControllerHostName(){
+	DWORD dwRet;
+	PDOMAIN_CONTROLLER_INFO pdcInfo;
+
+	dwRet = NETAPI32$DsGetDcNameA(NULL, NULL, NULL, NULL, 0, &pdcInfo);
+	if (ERROR_SUCCESS == dwRet) {
+		internal_printf("[*] Using DC: %s\n", pdcInfo->DomainControllerName);		
+	} else {
+ 	    BeaconPrintf(CALLBACK_ERROR, "Failed to identify PDC, are we domain joined?");
+	}
+
+	PCHAR hostName = pdcInfo->DomainControllerName + 2; // Remove preceeding backslashes
+
+	NETAPI32$NetApiBufferFree(pdcInfo);
+	return hostName;
+}
+
+LDAP* InitialiseLDAPConnection(PCHAR hostName, PCHAR distinguishedName){
+	LDAP* pLdapConnection = NULL;
+
+    pLdapConnection = WLDAP32$ldap_init(hostName, 389);
+    
+    if (pLdapConnection == NULL)
+    {
+      	BeaconPrintf(CALLBACK_ERROR, "Failed to establish LDAP connection on 389.");
+        WLDAP32$ldap_unbind(pLdapConnection);
```

Comment:
if pLdapConnection is null, we do not want to attempt to call unbind on it.

we also want to have a clean exit path after any of these failures, the code body itself should not continue to operate after a failure like this has occurred.